geo/src/algorithm/line_intersection.rs

```diff
@@ -0,0 +1,666 @@
+        (Collinear, Collinear, Collinear, Collinear)
+    ) {
+        return collinear_intersection(p, q);
+    }
+
+    // At this point we know that there is a single intersection point (since the lines are not
+    // collinear).
+    //
+    // Check if the intersection is an endpoint. If it is, copy the endpoint as the
+    // intersection point. Copying the point rather than computing it ensures the point has the
+    // exact value, which is important for robustness. It is sufficient to simply check for an
+    // endpoint which is on the other line, since at this point we know that the inputLines
+    // must intersect.
+    if p_q1 == Collinear || p_q2 == Collinear || q_p1 == Collinear || q_p2 == Collinear {
+        // Check for two equal endpoints.
+        // This is done explicitly rather than by the orientation tests below in order to improve
+        // robustness.
+        //
+        // [An example where the orientation tests fail to be consistent is the following (where
+        // the true intersection is at the shared endpoint
+        // POINT (19.850257749638203 46.29709338043669)
+        //
+        // LINESTRING ( 19.850257749638203 46.29709338043669, 20.31970698357233 46.76654261437082 )
+        // and
+        // LINESTRING ( -48.51001596420236 -22.063180333403878, 19.850257749638203 46.29709338043669 )
+        //
+        // which used to produce the INCORRECT result: (20.31970698357233, 46.76654261437082, NaN)
+
+        let intersection: Coordinate<F>;
+        // false positives for this overzealous clippy https://github.com/rust-lang/rust-clippy/issues/6747
+        #[allow(clippy::suspicious_operation_groupings)]
+        if p.start == q.start || p.start == q.end {
+            intersection = p.start;
+        } else if p.end == q.start || p.end == q.end {
+            intersection = p.end;
+            // Now check to see if any endpoint lies on the interior of the other segment.
+        } else if p_q1 == Collinear {
+            intersection = q.start;
+        } else if p_q2 == Collinear {
+            intersection = q.end;
+        } else if q_p1 == Collinear {
+            intersection = p.start;
+        } else {
+            assert_eq!(q_p2, Collinear);
+            intersection = p.end;
+        }
+        Some(LineIntersection::SinglePoint {
+            intersection,
+            is_proper: false,
+        })
+    } else {
+        let intersection = proper_intersection(p, q);
+        Some(LineIntersection::SinglePoint {
+            intersection,
+            is_proper: true,
+        })
+    }
+}
+
+fn collinear_intersection<F: GeoFloat>(p: Line<F>, q: Line<F>) -> Option<LineIntersection<F>> {
+    let p_rect = p.bounding_rect();
+    let p1q1p2 = p_rect.intersects(&q.start);
+    let p1q2p2 = p_rect.intersects(&q.end);
+
+    let q_rect = q.bounding_rect();
+    let q1p1q2 = q_rect.intersects(&p.start);
+    let q1p2q2 = q_rect.intersects(&p.end);
+
+    if p1q1p2 && p1q2p2 {
```

Comment:
This check could go before the `q_rect` assignment above, if we wanted to save a few calls. Breaks the flow a little so it's a tradeoff  🤷🏻